### PR TITLE
Rework static/shared variants into a configure-time build-switch (#688)

### LINKED ISSUE #826: Conda build problems for 9.3-backports on windows
It compiled but linking failed:
LINK : fatal error LNK1181: cannot open input file 'msvcstub.lib'

Where does this lib comes from?
set(OMNIORB_PKG_LIBRARIES_DYN "omniORB4_rt.lib;omniDynamic4_rt.lib;omnithread_rt.lib;COS4_rt.lib;msvcstub.lib")

Is it supposed to come with omniORB?
It looks like it was removed in tango-9-lts.

/Benjamin

﻿On 2021-01-15, 15:12, "Thomas Braun" <thomas.braun@byte-physics.de> wrote:

    Am 1/15/2021 um 2:16 PM schrieb Benjamin Bertrand:

    Hi Benjamin,

    the mentioned PR has already been merged. So if you use 0abc6ccc (Merge 
    pull request #824 from tango-controls/no-sed-backport, 2021-01-15) 
    instead of the 9.3.4 tag it should already work.

    Thomas

    > Hi Thomas,
    > 
    > Thanks. Indeed, no sed as there is no Cygwin.
    > Looks like it should be easy to apply the change from this commit as a patch.
    > I'll give it a try and let you know.
    > 
    > /Benjamin
    > 
    > On 2021-01-14, 23:43, "Thomas Braun" <thomas.braun@byte-physics.de> wrote:
    > 
    >      Am 1/14/2021 um 10:36 PM schrieb Benjamin Bertrand:
    > 
    >      Hi Benjamin,
    > 
    >      thx.
    > 
    >      The error from
    > 
    >      ...
    > 
    >      %SRC_DIR%\cppapi\server\device_3.cpp(926): error C2039:
    >      'set_attr_mutex': is not a member of 'Tango::AttributeValue_5'
    >      %SRC_DIR%\cppapi\server\idl/tango.h(3395): note: see declaration of
    >      'Tango::AttributeValue_5'
    >      %SRC_DIR%\cppapi\server\device_3.cpp(928): error C2039:
    >      'set_attr_mutex': is not a member of 'Tango::AttributeValue_5'
    >      %SRC_DIR%\cppapi\server\idl/tango.h(3395): note: see declaration of
    >      'Tango::AttributeValue_5'
    >      %SRC_DIR%\cppapi\server\device_3.cpp(940): error C2039:
    >      'set_attr_mutex': is not a member of 'Tango::AttributeValue_4'
    >      %SRC_DIR%\cppapi\server\idl/tango.h(3363): note: see declaration of
    >      'Tango::AttributeValue_4'
    >      %SRC_DIR%\cppapi\server\device_3.cpp(942): error C2039:
    >      'set_attr_mutex': is not a member of 'Tango::AttributeValue_4'
    >      %SRC_DIR%\cppapi\server\idl/tango.h(3363): note: see declaration of
    >      'Tango::AttributeValue_4'
    >      %SRC_DIR%\cppapi\server\device_3.cpp(959): error C2039:
    >      'rel_attr_mutex': is not a member of 'Tango::AttributeValue_5'
    >      %SRC_DIR%\cppapi\server\idl/tango.h(3395): note: see declaration of
    >      'Tango::AttributeValue_5'
    >      %SRC_DIR%\cppapi\server\device_3.cpp(968): error C2039:
    >      'rel_attr_mutex': is not a member of 'Tango::AttributeValue_4'
    >      %SRC_DIR%\cppapi\server\idl/tango.h(3363): note: see declaration of
    >      'Tango::AttributeValue_4'
    > 
    >      ...
    > 
    >      means that patching the code generated from omniidl was not done. The
    >      code at
    >      https://github.com/tango-controls/cppTango/blob/f4b4b652a90ffd82d2ecf7bb4da79a0f5d783187/cppapi/server/idl/CMakeLists.txt#L31
    >      should do that. I would guess that sed is missing.
    > 
    >      At some point we had a PR which fixed that in tango-9-lts, see
    >      https://github.com/tango-controls/cppTango/pull/638, but looks like this
    >      was not backported to 9.3-backports.
    > 
    >      Thomas
    > 
    >      > Hi!
    >      >
    >      > Please find attached the windows failed conda build log.
    >      >
    >      > I’ll try to push my branch somewhere if you want to try locally.
    >      >
    >      > /Benjamin

### PR DESCRIPTION
**Pull request by t-b**
Shared/static switching on windows

configure/CMakeLists.txt: Don't link against msvcstub anymore on windows

We don't link against that library from omniorb anymore as it is only needed as a workaround for using ominORB DLLs under MSVC++ 5.0 (due to destructor inlining issues).

Co-authored-by: Marius Elvert <marius.elvert@softwareschneiderei.de>

Close #826

### CODE REVIEW

**Comment by beenje** [comment]
I see that CI failed but I managed to build this branch on windows using conda-build (with only a patch to rename the libzmq and pthread libraries according to existing conda packages). I didn't run any tests.

**Comment by t-b** [comment]
@beenje Thanks for trying. I think I fixed the build error. It was in appveyor.yml.

**Comment by beenje** [comment]
> @beenje Thanks for trying. I think I fixed the build error. It was in appveyor.yml.

OK. If that's the only file you modified, that will not change anything for the conda build. I can still double check when the CI passes if you want.

**Comment by t-b** [comment]
@mliszcz @bourtemb Ready.

File: configure/cpack_win.cmake [hunk 1]
```diff
@@ -1,33 +1,20 @@
+set(COMPILER_TAG ${CMAKE_VS_PLATFORM_TOOLSET})
```
Comment (mliszcz) [inline configure/cpack_win.cmake:1]: For some reason `CMAKE_VS_PLATFORM_TOOLSET` was not used with `MSVC90`, but now it is. Are you sure it is defined? The documentation only mentions that: *VS 10 and above use MSBuild under the hood ...*

https://cmake.org/cmake/help/latest/variable/CMAKE_VS_PLATFORM_TOOLSET.html

Comment (mliszcz) [reply configure/cpack_win.cmake:1]: Yes, it is undefined. Just see the logs:
```
CPack: - package: C:/projects/static_build/libtango_9.3.4__x64_static_release.msi generated
```

Other jobs print version here:
```
CPack: - package: C:/projects/static_build/libtango_9.3.4_v141_x64_static_debug.msi generated.
```

Comment (t-b) [reply configure/cpack_win.cmake:1]: Thanks for carerful checking. I'll find a workaround.

Comment (t-b) [reply configure/cpack_win.cmake:1]: @mliszcz Done.

**Review by mliszcz — APPROVED** [review]
Looks good now! First msvc9 job just completed with correct package name:
```
CPack: - package: C:/projects/static_build/libtango_9.3.4_v90_x86_static_release.msi generated.
```

One comment though: CMAKE_VS_PLATFORM_TOOLSET variable is [supposed to be set by CMake](https://cmake.org/cmake/help/latest/manual/cmake-variables.7.html#variables-that-provide-information) to give use some information, not the other way around. I wouldn't overwrite that one but rather COMPILER_TAG which is owned by us.
But since the current solution works, and it's for 9.3 and for msvc9 only, I think it won't do any harm.

**Pull request closed without merging**